pkg/cache/cache.go

```diff
@@ -0,0 +1,275 @@
+			glog.Debugf("%#v and %#v do not match.Updating it in calico cache.", newObj, existingObj)
+
+			c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+			c.workqueue.Add(key)
+		}
+	} else {
+		glog.Debugf("%#v not found in calico cache. Adding it in calico cache", newObj)
+
+		c.threadSafeCache.Set(key, newObj, cache.NoExpiration)
+		c.workqueue.Add(key)
+	}
+}
+
+func (c *calicoCache) Delete(key string) {
+
+	glog.Debug("Deleting %s in calico", key)
+	c.threadSafeCache.Delete(key)
+	c.workqueue.Add(key)
+}
+
+func (c *calicoCache) Get(key string) (interface{}, bool) {
+
+	obj, found := c.threadSafeCache.Get(key)
+	if found {
+		return obj, true
+	}
+	return nil, false
+}
+
+// Prime funtion adds object to threadSafeCache.
+// Only difference with Set() call is it does not queue the key
+// to workqueue. Stores pointer of value object in  cache
+func (c *calicoCache) Prime(key string, value interface{}) {
+
+	c.threadSafeCache.Set(key, &value, cache.NoExpiration)
+}
+
+// ListKeys returns list of calico cache keys in the form
+// of array of strings. Cache stores name field of objects as key.
+func (c *calicoCache) ListKeys() []string {
+
+	cacheItems := c.threadSafeCache.Items()
+	keys := make([]string, 0, len(cacheItems))
+	for k := range cacheItems {
+		keys = append(keys, k)
+	}
+
+	return keys
+}
+
+func (c *calicoCache) GetQueue() workqueue.RateLimitingInterface {
+	return c.workqueue
+}
+
+// Load all the calico datastore objects at the begining of run
+func (c *calicoCache) Run(stopChan chan struct{}, reconcilerPeriod string) {
+
+	// Retry priming of cache if connection to ETCD datastore fails
+	for err := c.primeCache(); err != nil; {
+  		glog.WithError(err).Errorf("Failed to prime Calico cache, retrying")
+	}
+
+	go c.reconcile(stopChan, reconcilerPeriod)
+}
+
+// primeCache() function populates Calico Cache with only calico objects
+// that are created by policy controller.
+func (c *calicoCache) primeCache() error {
+
+	etcdObjList, err := c.ListFunc()
+
+	if err != nil {
+		glog.Error(err)
+		return err
+	}
+
+	for _, profile := range etcdObjList {
+		calicoKey := c.KeyFunc(profile)
+		c.Prime(calicoKey, profile)
+	}
+
+	return nil
+}
+
+// reconcile() function runs every `reconcilerPeriod` and brings ETCD datastore
+// in sync with Calico cache. This is to correct any manual changes done by 
+// user in Calico ETCD.
+func (c *calicoCache) reconcile(stopChan chan struct{}, reconcilerPeriod string) {
+
+	duration, err := time.ParseDuration(reconcilerPeriod)
+	if (err!=nil){
+		glog.Fatalf("Invalid time duration format %s for reconciler. Some correct examples, 5m, 30s, 2m30s etc. Reconciler will not sync any objects.",reconcilerPeriod)
+		return
+	}
+
+	// If user has set duration to 0 then disable the reconciler job.
+	if(duration.Nanoseconds() == 0){
+		glog.Infof("Reconciler period set to %d. Disabling reconciler.", duration.Nanoseconds())
+		return
+	}
+
+	ticker := time.NewTicker(duration)
+	go func() {
```

Comment:
Going back to simple for loop implementation to reduce unnecessary complexity added by ticker.